CefSharp.Core/Internals/ClientAdapter.cpp

```diff
@@ -323,15 +350,36 @@ namespace CefSharp
 
         bool ClientAdapter::OnBeforeBrowse(CefRefPtr<CefBrowser> browser, CefRefPtr<CefFrame> frame, CefRefPtr<CefRequest> request, bool isRedirect)
         {
-            IRequestHandler^ handler = _browserControl->RequestHandler;
-            if (handler == nullptr)
+            if (browser->IsPopup())
             {
-                return false;
+                auto popupHandler = _browserControl->PopupHandler;
+                if (popupHandler != nullptr)
+                {
+                    auto browserId = browser->GetIdentifier();
+                    IBrowser^ entry;
+                    if (_popupBrowsers->TryGetValue(browserId, entry))
+                    {
+                        auto frameWrapper = gcnew CefFrameWrapper(frame, _browserAdapter);
+                        CefRequestWrapper^ wrapper = gcnew CefRequestWrapper(request);
+                        return popupHandler->OnBeforeBrowse(_browserControl, entry, wrapper, isRedirect, gcnew CefFrameWrapper(frame, _browserAdapter));
```

Comment:
> Do we get notifications when frames/browsers get destroyed?

For `CefBrowser` it's pretty clear and I think that's pretty straight forward to reuse the wrapper. For `CefFrame`, I'm not sure that there is such a notification.

> Have an internal .Net friendly CefFrameWrapper that we cache, and is fronted by a weak reference owning ClientCefFrameWrapper like object. i.e. ClientCefFrameWrapepr would contain a GC weak reference to our CefFrameWrapper instance.

What does the extra layer get us in this instance? If we Have a reference to the wrapper and simply release the `unamanged` resource, is there anything more required? (I'm not opposed to the idea, I'm just unsure what the extra overhead gets us)

> I'm not sure OnLoadingStateChange is the correct context for disposing of the instances

I'm not sure either, it was just a though in regards to the `Frame Wrapper`. Like when the next page had finished loading we dispose of all the previous frame instances. This might end up being one of those that's left to the user to properly manage (which is bound to cause leaks). The `async` callbacks are equally as problematic.

I'm totally open to suggestions, I know what your trying to achieve, and it makes sense :smile: Still no clearer on how we'd go about it.